android/engine/src/main/java/org/smartregister/fhircore/engine/task/WorkflowCarePlanGenerator.kt

```diff
@@ -0,0 +1,329 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.smartregister.fhircore.engine.task
+
+import android.content.Context
+import ca.uhn.fhir.context.FhirContext
+import ca.uhn.fhir.context.FhirVersionEnum
+import com.google.android.fhir.FhirEngine
+import com.google.android.fhir.knowledge.KnowledgeManager
+import com.google.android.fhir.search.search
+import com.google.android.fhir.workflow.FhirOperator
+import dagger.hilt.android.qualifiers.ApplicationContext
+import java.io.File
+import java.time.Instant
+import java.util.Date
+import javax.inject.Inject
+import javax.inject.Singleton
+import org.hl7.fhir.r4.model.Bundle
+import org.hl7.fhir.r4.model.CanonicalType
+import org.hl7.fhir.r4.model.CarePlan
+import org.hl7.fhir.r4.model.IdType
+import org.hl7.fhir.r4.model.Library
+import org.hl7.fhir.r4.model.MetadataResource
+import org.hl7.fhir.r4.model.Patient
+import org.hl7.fhir.r4.model.PlanDefinition
+import org.hl7.fhir.r4.model.Reference
+import org.hl7.fhir.r4.model.Resource
+import org.hl7.fhir.r4.model.ResourceType
+import org.hl7.fhir.r4.model.Task
+
+@Singleton
+class WorkflowCarePlanGenerator
+@Inject
+constructor(
+  val knowledgeManager: KnowledgeManager,
+  val fhirOperator: FhirOperator,
+  val fhirEngine: FhirEngine,
+  @ApplicationContext val context: Context,
+) {
+
+  private var cqlLibraryIdList = ArrayList<String>()
+  private val jsonParser = FhirContext.forCached(FhirVersionEnum.R4).newJsonParser()
+
+  private fun writeToFile(resource: Resource): File {
+    val fileName =
+      if (resource is MetadataResource && resource.name != null) {
+        resource.name
+      } else {
+        resource.idElement.idPart
+      }
+    return File(context.filesDir, fileName).apply {
+      writeText(jsonParser.encodeResourceToString(resource))
+    }
+  }
+
+  /**
+   * Extracts resources present in PlanDefinition.contained field
+   *
+   * We cannot use $data-requirements on the [PlanDefinition] yet. So, we assume that all knowledge
+   * resources required to $apply a [PlanDefinition] are present within `PlanDefinition.contained`
+   *
+   * @param planDefinition PlanDefinition resource for which dependent resources are extracted
+   */
+  suspend fun getPlanDefinitionDependentResources(
+    planDefinition: PlanDefinition,
+  ): Collection<Resource> {
+    var bundleCollection: Collection<Resource> = mutableListOf()
+
+    for (resource in planDefinition.contained) {
```

Comment:
the contained resources might be handled by PlanProcessor within code, however, the resources which are not in contained like ActivityDefintion in WHO plans, Library referenced by PlanDef might need to be part of KnowledgeManager. Kindly double check.